resource/graph_node.go

```diff
@@ -481,6 +482,12 @@ func (w *GraphNode) canTransitionTo(state NodeState) bool {
 			return true
 		}
 	case NodeStateRemoving:
+	case NodeStateUnhealthy:
+		//nolint
+		switch state {
+		case NodeStateReady, NodeStateRemoving, NodeStateUnhealthy:
```

Comment:
When would a node transition from `Unhealthy` directly to `Ready`? Wouldn't it need to pass through `Configuring` first?

Reply:
see https://github.com/viamrobotics/rdk/pull/4257/files#r1702215112

I admit this is a bit of a shortcut, but right now `unhealthy` means a resource both encountered an error and needs reconfiguration. This works because a resource can only become unhealthy (e.g. encounter an error) during (re)configuration. This is how also it worked before we added states - resources with errors got fixed because their `needsConfiguration` flag was never cleared, so they would be processed by reconfigure.

I initially tried to force a the state transition that you are suggesting (e.g. an `unhealthy` resource MUST go back to `configuring` before it can be `ready`), but this is currently tricky because of how resources disappear from the the resource list once they become unhealthy.

Reply:
Thanks for the explanation!

> but this is currently tricky because of how resources disappear from the the resource list once they become unhealthy.

To me that sounds like `Unhealthy` -> `Removing` -> `no state because not in graph` -> `Configuring` -> `Ready`. Or does the graph node stick around and only move between `Unhealthy` -> `Ready`?

Reply:
A nudge that I'd like this detail/backstory/description of limitations that informed the written code be added as code documentation.

Reply:
> To me that sounds like Unhealthy -> Removing -> no state because not in graph -> Configuring -> Ready. Or does the graph node stick around and only move between Unhealthy -> Ready?

The latter is more accurate I think - technically the node sticks around but is not visible to most reconfiguration operations since they rely on the [`GraphNode.Resource` method](https://github.com/viamrobotics/rdk/blob/39ab463dd068b45beec73d44cab04c3a473c45c0/resource/graph_node.go#L137). Once this node successfully configured it gets inserted into the existing but hidden node and is marked as ready.

> A nudge that I'd like this detail/backstory/description of limitations that informed the written code be added as code documentation.

Absolutely